### PR TITLE
Add git hooks

### PR DESCRIPTION
**Pull request by glennmatthews**
# Installation and Updating
- When you first clone the repo, you need to manually run `./git/bin/update-hooks` to install the hooks into your `.git/hooks/` directory (Git doesn't let you commit `.git/hooks/`, this directory is local only).
- If you locally edit any of the files in `git_hooks/`, you can manually re-run `./git/bin/update-hooks` as needed.
- Once the hooks are installed, any time you do a `git rebase` or `git merge`, `update-hooks` will automatically be re-run and if there are any changes in `git_hooks/` as a result of the pull/merge/rebase, your local hooks will be automatically updated.
# Hooks
## pre-commit

Based on the pre-commit.sample script provided by Git itself, this just does a couple of simple checks:
- Don't commit files with non-ASCII filenames
- Don't commit any trailing whitespace
## pre-push

On the reasoning that you may commit locally several times before your code is 'ready' to push, this is the more heavy-weight hook:
- Don't push if RuboCop is failing
- Don't push if you forgot to update CHANGELOG.md
## post-rewrite and post-merge
- These just automatically update your locally installed hooks after a git pull/rebase/merge operation.
## post-flow-release-start
- When you do a 'git flow release start x.y.z', this does:
  - Replace the 'Unreleased' header in CHANGELOG.md with the given version tag and automatically adds the corresponding GitHub version comparison link.
  - Update version.rb with the specified version number
## post-flow-hotfix-start
- 'git flow hotfix start x.y.z'
- Very similar to post-flow-release-start, but slightly different in the handling of CHANGELOG.md since there should never be an 'Unreleased' header in the master branch.
## post-flow-release-finish
- When you do a 'git flow release finish x.y.z', this does:
  - Insert a new 'Unreleased' header in CHANGELOG.md for the develop branch
  - Bump the bugfix version (x.y.z -> x.y.z+1) and mark version.rb in the develop branch as 'x.y.zz-dev')

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -8,7 +8,8 @@ Cisco Network Elements support a rich set of features to make networks robust, e
 
 ## Making Changes
 
-* Fork the repository
+* Fork and clone the repository
+  * Run the `bin/git/update-hooks` script to install our recommended Git hooks into your local repository.
```
Comment (mikewiebe) [inline CONTRIBUTING.md:12]: Just curious.. if we don't run the install hooks then these checks are not enforced right?

Comment (glennmatthews) [inline CONTRIBUTING.md:12]: Correct. All of these are client-side hooks (github doesn't permit server-side hooks) so all they can do is assist with best practices, not force anything absolutely.

**Comment by robert-w-gries** [comment]
Suggestions:

# pre-commit
- All rubocop --lint checks should pass before commit
- [Sanitize commit messages](http://chris.beams.io/posts/git-commit/#seven-rules)
  - Separate subject from body with a blank line
  - Limit the subject line to 50 characters
  - Capitalize the subject line
  - Do not end the subject line with a period
  - Wrap the body at 72 characters

# pre-push
- Make sure your new minitest file(s) are included in tests/test_all_cisco.rb
- Do not allow pushes to master

# pre-rebase
- Do not allow rebase on remote branches
- Do not allow rebase on code previously pushed to remote repo

**Comment by glennmatthews** [comment]
```
All rubocop --lint checks should pass before commit
```

I made this a pre-push check instead - I'm thinking people may want to have a local workflow that looks something like:
- Get the code 'working' and commit it (but do not push)
- Fix rubocop warnings, retest, and commit (but do not push)
- Squash the commits together and push

Basically I don't want to prevent people from using commit as a way to locally checkpoint their code.

**Comment by robert-w-gries** [comment]
From Rubocop's readme:

> Lint cops check for possible errors and very bad practices in your code.

If code fails a lint check, there's a good chance it's subtly dysfunctional code.  Catching errors like that pre-commit could save some effort.

**Comment by glennmatthews** [comment]
Ah I see your point. Require (just) lint before commit and a full run before push. Yeah, I can do that.

**Comment by robert-w-gries** [comment]
:+1: 

Very low priority suggestion: spell check for commit message in pre-commit

**Comment by mikewiebe** [comment]
Want to spend more time reviewing this, but don't want to hold up tooling to improve our workflow.

:+1:

**Comment by glennmatthews** [comment]
```
 > git commit --amend
*** Running Git hook script: pre-commit-rubocop...
warning: parser/current is loading parser/ruby21, which recognizes
warning: 2.1.7-compliant syntax, but you are running 2.1.1.
Inspecting 57 files
.........................................................

57 files inspected, no offenses detected
        ...[ OK ]
*** Running Git hook script: pre-commit-validate-diffs...
0
        ...[ OK ]
*** Running Git hook script: commit-msg-enforce_style...
        ...[ OK ]
*** Running Git hook script: post-rewrite-update-hooks...
        ...[ OK ]
[feature/git_hooks 657f0f3] Modularize hooks ; validate commit messages
<snip>
```

```
> git push --force
*** Running Git hook script: pre-push-check-changelog...
        Checking CHANGELOG.md...
        ...[ OK ]
*** Running Git hook script: pre-push-rubocop...
warning: parser/current is loading parser/ruby21, which recognizes
warning: 2.1.7-compliant syntax, but you are running 2.1.1.
Inspecting 57 files
.........................................................

57 files inspected, no offenses detected
        ...[ OK ]
Counting objects: 8, done.
<snip>
```

**Pull request merged**